crates/edr_evm/src/state/trie/state_trie.rs

```diff
@@ -0,0 +1,98 @@
+use std::sync::Arc;
+
+use alloy_rlp::Decodable;
+use edr_eth::account::BasicAccount;
+
+use crate::{
+    state::trie::{persistent_memory_db::PersistentMemoryDB, trie_query::TrieQuery},
+    AccountInfo, Address, B256,
+};
+
+#[derive(Debug)]
+pub(super) struct StateTrie {
+    db: Arc<PersistentMemoryDB>,
+    root: B256,
+}
+
+impl<'a> StateTrie {
+    pub fn account(&self, address: &Address) -> Option<BasicAccount> {
+        self.trie_query().get(address).map(|encoded_account| {
+            BasicAccount::decode(&mut encoded_account.as_slice()).expect("Valid RLP")
+        })
+    }
+
+    /// Create a helper struct that allows setting and removing multiple
+    /// accounts and then updates the state root when dropped.
+    pub fn mutate(&'a mut self) -> StateTrieMutation<'a> {
+        let trie_query = self.trie_query();
+        StateTrieMutation {
+            state_trie: self,
+            trie_query,
+        }
+    }
+
+    pub fn root(&self) -> B256 {
+        self.root
+    }
+
+    pub fn trie_query(&self) -> TrieQuery {
+        TrieQuery::new(Arc::clone(&self.db), &self.root)
+    }
+}
+
+impl Clone for StateTrie {
+    fn clone(&self) -> Self {
+        Self {
+            db: Arc::new((*self.db).clone()),
+            root: self.root,
+        }
+    }
+}
+
+impl Default for StateTrie {
+    fn default() -> Self {
+        let db = Arc::new(PersistentMemoryDB::default());
+        let mut trie = TrieQuery::empty(Arc::clone(&db));
+        let root = trie.root();
+
+        // TODO root should be cacheable
```

Comment:
Should this be handled in this PR? If not, could you please create a GitHub issue for it and link it?